cmd/selfupdate/selfupdate.go

```diff
@@ -0,0 +1,327 @@
+		if err != nil {
+			return err
+		}
+		fs.Debugf(nil, "downloaded hashsum list: %s", hashsumsURL)
+
+		wantHash, err := findFileHash(hashsumsBuf, archiveFilename)
+		if err != nil {
+			return err
+		}
+
+		if !bytes.Equal(wantHash, gotHash[:]) {
+			return fmt.Errorf("hash mismatch: want %02x vs got %02x", wantHash, gotHash)
+		}
+	}
+
+	entryName := fmt.Sprintf("rclone-%s-%s-%s/rclone", version, runtime.GOOS, runtime.GOARCH)
+	if runtime.GOOS == "windows" {
+		entryName += ".exe"
+	}
+
+	// Extract executable to a temporary file, then replace it by an instant rename
+	err = extractZipToFile(archiveBuf, entryName, newFile)
+	if err != nil {
+		return err
+	}
+	fs.Debugf(nil, "extracted %s to %s", entryName, newFile)
+	return nil
+}
+
+func verifyAccess(file string) error {
+	admin := "root"
+	if runtime.GOOS == "windows" {
+		admin = "Administrator"
+	}
+
+	fileInfo, fileErr := os.Lstat(file)
+
+	if fileErr != nil {
+		dir := filepath.Dir(file)
+		dirInfo, dirErr := os.Lstat(dir)
+		if dirErr != nil {
+			return dirErr
+		}
+		if !dirInfo.Mode().IsDir() {
+			return fmt.Errorf("%s: parent path is not a directory, specify a different path using --output", dir)
+		}
+		if !writable(dir) {
+			return fmt.Errorf("%s: directory is not writable, please run self-update as %s", dir, admin)
+		}
+	}
+
+	if fileErr == nil && !fileInfo.Mode().IsRegular() {
+		return fmt.Errorf("%s: path is not a normal file, specify a different path using --output", file)
+	}
+
+	if fileErr == nil && !writable(file) {
+		return fmt.Errorf("%s: file is not writable, run self-update as %s", file, admin)
+	}
+
+	return nil
+}
+
+func findFileHash(buf []byte, filename string) (hash []byte, err error) {
+	lines := bufio.NewScanner(bytes.NewReader(buf))
+	for lines.Scan() {
+		tokens := strings.Split(lines.Text(), "  ")
+		if len(tokens) == 2 && tokens[1] == filename {
+			if hash, err := hex.DecodeString(tokens[0]); err == nil {
+				return hash, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("%s: unable to find hash", filename)
+}
+
+func extractZipToFile(buf []byte, entryName, newFile string) error {
+	zipReader, err := zip.NewReader(bytes.NewReader(buf), int64(len(buf)))
+	if err != nil {
+		return err
+	}
+
+	var reader io.ReadCloser
+	for _, entry := range zipReader.File {
+		if entry.Name == entryName {
+			reader, err = entry.Open()
+			break
+		}
+	}
+	if reader == nil || err != nil {
+		return fmt.Errorf("%s: file not found in archive", entryName)
+	}
+	defer func() {
+		_ = reader.Close()
+	}()
+
+	err = os.Remove(newFile)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("%s: unable to create new file: %v", newFile, err)
+	}
+	writer, err := os.OpenFile(newFile, os.O_CREATE|os.O_EXCL|os.O_WRONLY, os.FileMode(0755))
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(writer, reader)
+	_ = writer.Close()
+	if err != nil {
+		_ = os.Remove(newFile)
+	}
+	return err
+}
+
+func downloadFile(url string) ([]byte, error) {
```

Comment:
Downloading the release in memory seems OK provided the release doesn't get too big!

Reply:
Temporary file? Doable.
But I didn't see rclone releases more than 50-60M
It's not big even for embedded boxes

Reply:
What threshold do you think is reasonable?

Reply:
I think we should leave your code as-is and not worry about it for the moment.

```
$ rclone --fast-list lsl memstore:downloads-rclone-org --include *.zip | sort -n | tail
 19355621 2019-10-05 12:17:21.161062944 v1.49.5/rclone-v1.49.5-osx-amd64.zip
 19624754 2019-11-19 16:21:16.063618980 v1.50.2/rclone-v1.50.2-osx-386.zip
 19625809 2019-11-02 15:03:25.089316344 v1.50.1/rclone-v1.50.1-osx-386.zip
 19633882 2019-10-26 11:55:15.169383338 v1.50.0/rclone-v1.50.0-osx-386.zip
 19914331 2020-02-01 10:59:48.391080948 v1.51.0/rclone-v1.51.0-osx-386.zip
 22953431 2020-06-24 14:08:44.602862189 v1.52.2/rclone-v1.52.2-osx-386.zip
 22956842 2020-08-08 09:44:20.717505397 rclone-current-osx-386.zip
 22956842 2020-08-08 09:44:20.717505397 v1.52.3/rclone-v1.52.3-osx-386.zip
 22970878 2020-05-27 17:53:06.774147983 v1.52.0/rclone-v1.52.0-osx-386.zip
 22981766 2020-06-10 16:27:10.754567269 v1.52.1/rclone-v1.52.1-osx-386.zip
```

The biggest official release is 23M which as you said is peanuts now-a-days.

Reply:
leaving as is